[feat] 상품 목록 조회 기능

pie2457 opened the pull request:
## Issues
- #20 

## What is this PR? 👓
- 상품 목록 조회에 대한 PR입니다.

## Key changes 🔑
- QueryDsl에 대한 의존성 추가
- 페이징을 사용한 상품 목록 조회 / 카테고리별 상품 목록 조회
## 📖 공부하면서 배웠던 것
### 📌 Pagination을 사용하는 이유
사용자가 애플리케이션을 사용 중 게시판, 상품 목록 등을 요청할 때 결과 값이 총 100만개일 경우 매번 전체를 전부 가져오게되면 매우 느려지며 사용자는 불편을 느끼게된다. 하지만 데이터를 조금씩 (20~100개) 나눠서 가져오고 사용자가 원하는 경우 다음 데이터를 가져오게 되면 훨씬 빠르고 사용자도 애플리케이션에 대해 만족할 것이다. 이러한 이유로 Pagination을 사용한다.
### Pagination 종류
- offset 방식 : offset과 limit 예약어를 통하여 select의 전체 결과 중 일부만 가져오는 방법이다.
- no-offset(=cursor) 방식 : cursor는 어떠한 row를 가르키는 포인터이고, 이 cursor가 가르키는 row로부터 일정 개수만큼 가져오는 방식이다.
### 📌 offset
**offset**이란 sql에서 조회를 시작할 **기준점을 의미**하고 **limit**는 **조회할 결과의 개수를 의미**한다. 예를 들어 아래와 같은 쿼리는
~~~sql
SELECT * FROM {테이블 이름} LIMIT 20 OFFSET 60;
~~~
60번째 행부터 20개의 행을 읽겠다는 의미이다. (이때 행은 0부터 시작.)
 offset은 조회를 한 결과에서 limit으로 지정한 개수만큼만 반환하고 나머지는 버리는 식으로 동작한다. 
위 쿼리로 예를 들자면 60번째 데이터 부터 20개를 조회하기 위해서 80개의 데이터를 모두 읽은 뒤, 앞의 필요없는 60개의 데이터는 버려야 한다. 
예시처럼 적은양의 데이터를 조회할 때는 성능적인 문제가 발생하지 않지만, 전체 데이터의 개수가 많아질수록 버려지는 데이터의 양이 많아져 문제가 된다. 
### offset 방식의 단점
- 뒷부분 쿼리 시 속도 저하
    
    위에서 말했듯이 offset방식은 여러개의 데이터를 한번에 가져와 필요한 만큼만 출력하고 나머지는 버리게 된다. 예를 들어 1,000,000개의 데이터를 가지고 있고, 100,000번째의 데이터에서 20개만 필요하다고 가정해보자. 내가 필요한 만큼은 20개인데 데이터는 100,020개를 가져오고 100,000개의 데이터는 버려지게 된다. 그 숫자는 커지면 커질수록 속도는 저하된다.
    
- 데이터 누락
    offset 방식은 데이터의 잦은 추가와 삭제가 있을 경우 데이터의 중복과 누락이 발생할 수 있다. offset 방식은 전체 데이터 중 offset만큼 건너 뛰며 이때 이전에 받은 데이터를 고려하지 않고 매번 새로 계산한다.
    - 예를 들어, 사용자가 1페이지의 1번부터 5번까지의 상품을 보고있는데 관리자가 4,5번 상품을 삭제했다고 가정시 사용자의 2페이지 요청시 데이터베이스는
    `나는 1번부터 5개를 줬는.. 어? 4번 5번이 없네? 그럼 7번까지 줬구나 !`
     라고 생각하고 2페이지는 8번 상품부터 5개의 데이터를 보내줄것이다. 그러면 6번상품과 7번상품의 데이터가 누락이 된다.

   이는 해당 서비스를 재접속하거나 다시 1페이지를 요청하지 않는 이상(새로고침) 다시는 볼 수 없게 된다.
- 데이터 중복
<img src="https://github.com/pie2457/Tomato-market/assets/104147789/b66bdd60-241f-4fba-8037-e7159cc44fac" width="700" height="140">

페이지를 이동하는 중간에 데이터가 삽입, 혹은 삭제 등의 작업이 발생한다면 다른 페이지에 영향을 주게 된다. 예를 들어, 1페이지에서 3개의 데이터를 조회한 후 2페이지로 넘어가는 과정에서 새로운 데이터가 추가된다면? 

<img src="https://github.com/pie2457/Tomato-market/assets/104147789/7beb34ea-ea67-4151-ad2c-496eaf0a0404" width="730" height="140">

이전 페이지에서 읽었던 데이터가 밀려와 이미 조회한 데이터를 중복 조회하게 된다.
### 📌 no-offset(=cursor)
**no-offset(=cursor)** 방식은 이전 페이지의 마지막 데이터의 id 값을 기억하고, 다음 페이지를 요청할 때 이 id 값 이후의 데이터를 가져오는 방식이다. 이를 통해 매번 offset 값을 계산하지 않아도 되기 때문에 데이터베이스에서 더욱 효율적으로 데이터를 가져올 수 있다. 이 방식을 사용할 경우 기준점 이전의 데이터도 모두 조회하던 offset과 달리 기준점(=cursor)부터 limit의 개수만 조회하기 때문에 데이터의 개수가 많아져도 성능 문제가 발생하지 않는다. 보통 무한스크롤이라 칭하기도 하고, sns에서 많이 사용되고 있다.
~~~sql
SELECT * FROM {테이블} 
WHERE {조건문}
AND id < lastId
ORDER BY id DESC 
LIMIT {컨텐츠개수}
~~~
예를 들어, 유저가 N번째 데이터부터 5개의 데이터를 조회하고 싶다고 요청한다면 직전에 조회한 데이터들 중 마지막 데이터의 id를 조건문으로 사용하여 이전에 조회했던 데이터들을 읽을 필요 없이 원하는 데이터로 건너 뛸 수 있게 된다. 때문에 페이지가 뒤로 가더라도 처음 페이지를 읽은 것과 동일한 성능을 가지게 된다. 
<img src="https://github.com/pie2457/Tomato-market/assets/104147789/20e68d79-76e5-4b33-a029-175be51235b2" width="600" height="300">
또한 페이지를 순차적으로 가져오기 때문에 중간에 삽입, 삭제, 업데이트 등의 작업이 발생해도 다른 페이지에 영향을 미치지 않는다. 이를 통해 데이터베이스의 부하를 줄일 수 있다.
아까와 동일한 예시로 첫 번째 페이지에서 3개의 데이터를 조회한 후 다음 페이지로 넘어가는 사이에 새로운 데이터가 추가되더라도 다음 페이지에 영향을 주지 않는다. (동일한 데이터를 중복 조회할 일이 없어진다.)
<img src="https://github.com/pie2457/Tomato-market/assets/104147789/6ae45077-3aba-4e73-9615-376b7f65ee9c" width="700" height="150">
<img src="https://github.com/pie2457/Tomato-market/assets/104147789/bfd935db-8f4d-41d5-94f6-d5fd25d87520" width="700" height="120">

pie2457 committed:
docs #20 : QueryDsl을 사용하기 위한 dependency 추가

pie2457 committed:
refactor #20 : region에 필드명 변경 및 추가

fullAddressName을 생각 못하다가 region 테이블에 데이터 넣다가 알아차려서 추가했습니다.

pie2457 committed:
feat #20 : 페이징 기능 구현

pie2457 committed:
feat #20 : 상품 목록 조회 기능 구현

pie2457 committed:
feat #20 : 상품 목록 조회 테스트 작성

23Yong commented on src/main/java/pie/tomato/tomatomarket/infrastructure/persistence/item/ItemPaginationRepository.java:
```diff
@@ -0,0 +1,45 @@
+				itemRepository.equalTradingRegion(region)
+			)
+			.orderBy(item.createdAt.desc())
+			.limit(size + 1)
```
왜 +1 해서 가져오셨나용~?

pie2457 replied:
다음 게시글이 있는지 확인하기 위해서 +1 해서 가져왔습니다 ~!

23Yong commented on src/main/java/pie/tomato/tomatomarket/presentation/request/item/ItemResponse.java:
```diff
@@ -0,0 +1,21 @@
+	private ItemStatus status;
+	private Long chatCount;
+	private Long wishCount;
+	private String isSeller;
```
`isSeller`가 `String` 타입인데 의도하신건가욤?
혹은 잘못된 필드가 들어가 있는 걸까요? 레포지토리에서는 sellerId를 가져오는 것 같던데요~!

pie2457 replied:
넹 의도한것입니다 어떤 판매자가 물건을 팔고있는지에 대한 닉네임을 반환하려고 한거였어요! 
sellerId는 제가 잘못 작성했네요 ㅠㅠ 수정하겠습니다 !

23Yong commented on src/test/java/pie/tomato/tomatomarket/application/ItemServiceTest.java:
```diff
@@ -91,23 +93,71 @@ void itemRegister() {
+		CustomSlice<ItemResponse> findItems = itemService.findAll(null, 10, null, null);
+
+		// then
+		assertThat(findItems.getContents().size()).isEqualTo(10);
```
사이즈가 10개인걸 테스트하는 것도 좋지만, `조회에 성공한다`의 기준에 맞춰서 테스트하면 좋을 것 같아요~!
예를 들어 현재 `조회에 성공한다`는 `사이즈`대로 가져오고 알맞은 `필드`들(title, content..)이 있고 생성시간 순으로 `정렬`이 되어있다. 라고 한다면 이를 테스트하면 좋을 것 같아용~!

pie2457 replied:
넵 그렇게 해서 수정해보겠습니다 !

23Yong commented on src/main/java/pie/tomato/tomatomarket/infrastructure/persistence/item/ItemRepository.java:
```diff
@@ -0,0 +1,52 @@
+
+	default SliceImpl<ItemResponse> checkLastPage(int size, List<ItemResponse> results) {
+
+		boolean hasNext = false;
```
default 메서드를 사용하셨군요~
`checkLastPage` 메서드는 itemRepository에 있는 것이 좋을까요? 현재는 item만을 페이징 고려했지만 region 까지 고려하면 동일한 메서드가 또 생길것 같아요! 그때는 regionRepository에 동일하게 results의 타입만 달라진 checkLastPage 메서드가 생기겠죠?

그래서 다른 레포지토리(ex. PaginationRepository)에 두고 다른 인터페이스가 이를 상속받고 이 메서드를 제너릭으로 생성하는 것도 괜찮을 것 같습니다.! 고려만 해봐주세요~

```java
default <T> SliceImpl<T> checkLastPage(int size, List<T> results)
```

pie2457 replied:
넵 수정하는게 더 좋을꺼같아요 ! 감사합니다 🥸~!

23Yong reviewed (approved):
수고하셨어요~

pie2457 committed:
refactor #20 : 리뷰사항 반영

- checkLastPage 메서드를 PaginationRepository로 분리
  - PaginationRepository 생성
- 상품 목록 조회 테스트 코드 수정
  - 기존 : 가져온 사이즈만 체크
  - 변경 : 사이즈 체크 및 객체의 필드가 알맞게 들어왔는지, 생성시간 순으로 정렬되어 있는지 등을 체크.
- ItemPaginationRepository에서 판매자의 닉네임을 isSeller 필드에 바인딩 해줘야하는데 생성되지 않은 필드 sellerId로 잘못가져와서 isSeller필드에는 데이터가 바인딩이 되지 않고 null이 들어옴 -> sellerId -> isSeller로 수정
